### PR TITLE
[FIRRTL][IMCP] Support Properties; support materializing string/bool.

### LINKED ISSUE #5940: [FIRRTL][IMCP] Don't crash on processing properties, getBaseType can be null
Input:
```
FIRRTL version 3.1.0
circuit PropertySpec_Anon :
  module _16_Anon :
    input propIn : Integer @[src/test/scala/chiselTests/properties/PropertySpec.scala 219:24]
    output propOut : Integer @[src/test/scala/chiselTests/properties/PropertySpec.scala 220:25]

    propassign propOut, propIn @[src/test/scala/chiselTests/properties/PropertySpec.scala 221:17]

  module PropertySpec_Anon :
    output propOut : Integer @[src/test/scala/chiselTests/properties/PropertySpec.scala 224:23]

    inst passthru of _16_Anon @[src/test/scala/chiselTests/properties/PropertySpec.scala 218:28]
    propassign passthru.propIn, Integer(123) @[src/test/scala/chiselTests/properties/PropertySpec.scala 223:23]
    propassign propOut, passthru.propOut @[src/test/scala/chiselTests/properties/PropertySpec.scala 225:15]
```

Crashes in IMCP, which expects getBaseType to never return null (it didn't previously, FWIW, but regardless).

This issue does not include the ask for IMCP to propagate property constants themselves, but to start with don't crash when encountering them.

### PR DESCRIPTION
**Pull request by dtzSiFive**
Fixes #5940 and supports basic property const prop while visiting.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 1]
```diff
@@ -392,15 +412,29 @@ LatticeValue IMConstPropPass::getExtendedLatticeValue(FieldRef value,
...
 
-  // If this is a BoolAttr then we are dealing with a special constant.
-  if (auto boolAttr = dyn_cast<BoolAttr>(constant)) {
-    // No extOrTrunc necessary for clock or reset types.
+  // No extOrTrunc necessary for bools or strings.
+  if (auto boolAttr = dyn_cast<BoolAttr>(constant))
     return LatticeValue(boolAttr);
-  }
+  if (auto strAttr = dyn_cast<StringAttr>(constant))
+    return LatticeValue(strAttr);
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:419]: maybe we can just return `result` for property type? `if(isa<PropertyType>()) return result;`

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp [hunk 2]
```diff
@@ -963,8 +1008,11 @@ void IMConstPropPass::rewriteModuleBody(FModuleOp module) {
       continue;
     }
 
-    // Don't "refold" constants.  TODO: Unique in the module entry block.
-    if (isa<ConstantOp, SpecialConstantOp, InvalidValueOp>(op))
+    // Don't "refold" constants.
+    // If you don't skip here, everything will crash mysteriously and
+    // inconsistently, and asan/ubsan cannot save you.  Here be dragons.
+    if (isa<ConstantOp, SpecialConstantOp, InvalidValueOp>(op) ||
+        op.hasTrait<mlir::OpTrait::ConstantLike>())
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:1015]: nit: maybe ConstantLike is sufficient here? ConstantOp and SpecialConstantOp have ConstantLike and InvalidValue is now overdefined.

**Review by uenoku — APPROVED** [review]
Extending IMCP to handle property looks very reasonable to me. LGTM

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 1 above)
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:419]: Oh! I somehow missed I could just return `result`, hence handling the supported types explicitly re:constructing LatticeValue for the right thing. Good call, thanks!

File: lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp (hunk 2 above)
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:1015]: Sounds good to me! I'm not sure checking the trait is perfectly legit but short of adding a thing that /does/ mean what we want ("will fold to an attr that will materialize to itself?") this seems right to me-- certainly communicates the intent :+1: .

**Pull request merged**